source/server/session_resource_repository.h

```diff
@@ -0,0 +1,226 @@
+  class ResourceMap {
+   public:
+    void add(uint32_t session_id, TResourceHandle handle);
+
+    bool contains(uint32_t session_id) const;
+    TResourceHandle get_handle(uint32_t session_id) const;
+    uint32_t get_session_id(TResourceHandle handle) const;
+
+    uint32_t remove_handle(TResourceHandle handle);
+    TResourceHandle remove_session_id(uint32_t session_id);
+
+   private:
+    using SessionToResourceMap = std::map<uint32_t, TResourceHandle>;
+    using ResourceToSessionMap = std::map<TResourceHandle, uint32_t>;
+    using MapLock = std::lock_guard<std::recursive_mutex>;
+
+    SessionToResourceMap map_;
+    ResourceToSessionMap reverse_map_;
+    mutable std::recursive_mutex map_mutex_;
+  };
+
+  SessionRepository* session_repository_;
+  ResourceMap resource_map_;
+};
+
+template <typename TResourceHandle>
+int SessionResourceRepository<TResourceHandle>::add_session(
+    const std::string& session_name,
+    InitFunc init_func,
+    CleanupSessionFunc cleanup_func,
+    uint32_t& session_id)
+{
+  uint32_t session_from_repository = 0;
+
+  auto session_creator = SessionResourceCreator(init_func);
+  auto status = session_repository_->add_session(
+      session_name,
+      [&]() { return session_creator.init_resource_handle(); },
+      // By val capture to keep cleanup_func in memory.
+      [=](uint32_t id) {
+        auto handle = resource_map_.remove_session_id(id);
+        return cleanup_func(handle);
+      },
+      session_from_repository);
+
+  if (status) {
+    session_id = 0;
+    return status;
+  }
+
+  if (session_creator.added_new_handle_) {
+    resource_map_.add(session_from_repository, session_creator.handle_);
+  }
+
+  // session_name resolves to a session in a different resource repository.
+  // Report error.
+  if (!resource_map_.contains(session_from_repository)) {
+    session_id = 0;
+    const int kError_InvalidArg = 1;
+    return kError_InvalidArg;
+  }
+
+  session_id = session_from_repository;
+  return 0;
+}
+
+template <typename TResourceHandle>
+TResourceHandle SessionResourceRepository<TResourceHandle>::access_session(uint32_t session_id, const std::string& session_name) const
+{
+  auto id = session_repository_->access_session(session_id, session_name);
+  return resource_map_.get_handle(id);
+}
+
+template <typename TResourceHandle>
+uint32_t SessionResourceRepository<TResourceHandle>::resolve_session_id(TResourceHandle handle) const
+{
+  return resource_map_.get_session_id(handle);
+}
+
+template <typename TResourceHandle>
+void SessionResourceRepository<TResourceHandle>::remove_session(TResourceHandle handle)
+{
+  auto id = resource_map_.remove_handle(handle);
+  if (id) {
+    session_repository_->remove_session(id);
+  }
+}
+
+template <typename TResourceHandle>
+uint32_t SessionResourceRepository<TResourceHandle>::SessionResourceCreator::init_resource_handle()
+{
+  auto init_result = init_func_();
+  auto status = std::get<0>(init_result);
```

Comment:
Do you need this anymore? The `init_func_()` will just return an `int32_t` now, right?

Reply:
Nevermind, marking this resolved. I was getting confused between the `InitFunc` typedef in this file and the one in the `session_repository`. This makes sense now.

Reply:
There are 2 different `InitFunc`s now.

On `SessionRepository`:  `typedef std::function<int32_t()> InitFunc;`
On `SessionResourceRepository`: `typedef std::function<std::tuple<int, TResourceHandle>()> InitFunc;`

This is the second one. The `InitFunc` passed in from the service still needs 2 return values because it creates the TResourceHandle (either a TaskHandle or a ViSession) and also returns a status. The `SessionResourceRepository` holds on to that and only returns the status to the `SessionRepository`.

Maybe just some renames would help with clarity?
Maybe `SessionResourceRepository::InitFunc` -> `SessionResourceRepository::CreateResourceFunc`?

Technically they both *do* the same thing, but they have different return values because of where they're invoked.

Reply:
Yeah, it was clear after I noticed there were two of these `InitFunc` functions defined. I had missed the one in defined in this file the first passthrough which led to the confusion.

If you want to do a rename I'd consider updating the cleanup function name as well so they're symmetrical like:
1. `InitFunc` and `CleanupFunc`
2. `CreateResourceFunc` and `CleanupResourceFunc`